patches/tModLoader/Terraria.ModLoader/NPCLoader.cs

```diff
@@ -946,6 +946,84 @@ private class HookList
 			}
 		}
 
+		private static HookList HookCanGoToKingStatue = AddHook<Func<NPC, bool?>>(g => g.CanGoToKingStatue);
+
+		public static bool CanGoToKingStatue(NPC npc) {
+			bool defaultCanGo = npc.modNPC?.CanGoToKingStatue() ?? false;
+
+			switch (npc.type) {
+				case NPCID.Merchant:
+				case NPCID.Guide:
+				case NPCID.ArmsDealer:
+				case NPCID.Demolitionist:
+				case NPCID.Clothier:
+				case NPCID.GoblinTinkerer:
+				case NPCID.Wizard:
+				case NPCID.SantaClaus:
+				case NPCID.Truffle:
+				case NPCID.DyeTrader:
+				case NPCID.Painter:
+				case NPCID.WitchDoctor:
+				case NPCID.Pirate:
+				case NPCID.LightningBug: // A literal bug
+				case NPCID.Angler:
+				case NPCID.DD2Bartender:
+					defaultCanGo = true;
+					break;
+			}
+
+			foreach (GlobalNPC g in HookCanGoToKingStatue.arr) {
+				bool? canGo = g.Instance(npc).CanGoToKingStatue(npc);
+				if (canGo.HasValue) {
+					if (!canGo.Value) {
+						return false;
+					}
+					defaultCanGo = true;
+				}
+			}
+
+			return defaultCanGo;
+		}
+
+		private static HookList HookCanGoToQueenStatue = AddHook<Func<NPC, bool?>>(g => g.CanGoToQueenStatue);
+
+		public static bool CanGoToQueenStatue(NPC npc) {
+			bool defaultCanGo = npc.modNPC?.CanGoToQueenStatue() ?? false;
+
+			switch (npc.type) {
+				case NPCID.Nurse:
+				case NPCID.Dryad:
+				case NPCID.Mechanic:
+				case NPCID.Steampunker:
+				case NPCID.PartyGirl:
+				case NPCID.Stylist:
+					defaultCanGo = true;
+					break;
+			}
+
+			foreach (GlobalNPC g in HookCanGoToQueenStatue.arr) {
+				bool? canGo = g.Instance(npc).CanGoToQueenStatue(npc);
+				if (canGo.HasValue) {
+					if (!canGo.Value) {
+						return false;
+					}
+					defaultCanGo = true;
```

Comment:
Just making sure why you set this to true. Is any NPC supposed to teleport if no Global class returns false? Earlier you are setting defaultCanGo to false if not a mod NPC and not any of the vanilla types.

Reply:
Any NPC is supposed to teleport if no Global class returns false and at least one of them returns true, and that's what the code is doing.

Reply:
Right, I forgot you return null otherwise.